sdk/keyvault/azure-security-keyvault-keys/src/cryptography/cryptography_client.cpp

```diff
@@ -0,0 +1,389 @@
+  for (uint64_t read = data.Read(buffer, DefaultStreamDigestReadSize); read > 0;
+       read = data.Read(buffer, DefaultStreamDigestReadSize))
+  {
+    hasAlgo->Append(buffer, static_cast<size_t>(read));
+  }
+  return hasAlgo->Final();
+}
+
+inline std::vector<uint8_t> CreateDigest(
+    SignatureAlgorithm algorithm,
+    std::vector<uint8_t> const& data)
+{
+  auto hasAlgo = algorithm.GetHashAlgorithm();
+  return hasAlgo->Final(data.data(), data.size());
+}
+} // namespace
+
+void CryptographyClient::Initialize(std::string const&, Azure::Core::Context const& context)
+{
+  if (m_provider != nullptr)
+  {
+    return;
+  }
+
+  try
+  {
+    auto key = m_remoteProvider->GetKey(context).Value;
+    m_provider = LocalCryptographyProviderFactory::Create(key);
+    if (m_provider == nullptr)
+    {
+      // KeyVaultKeyType is not supported locally. Use remote client.
+      m_provider = m_remoteProvider;
+      return;
+    }
+  }
+  catch (Azure::Core::RequestFailedException const& e)
+  {
+    if (e.StatusCode == HttpStatusCode::Forbidden)
+    {
+      m_provider = m_remoteProvider;
+    }
+    else
+    {
+      throw;
+    }
+  }
+  catch (...)
+  {
+    // re-throw any other exception
+    throw;
+  }
+}
+
+CryptographyClient::CryptographyClient(
+    std::string const& keyId,
+    std::shared_ptr<Core::Credentials::TokenCredential const> credential,
+    CryptographyClientOptions const& options,
+    bool forceRemote)
+{
+  auto apiVersion = options.Version.ToString();
+  m_keyId = keyId;
+  m_remoteProvider = std::make_shared<RemoteCryptographyClient>(keyId, credential, options);
+  m_pipeline = m_remoteProvider->Pipeline;
+
+  if (forceRemote)
+  {
+    m_provider = m_remoteProvider;
+  }
+}
+
+EncryptResult CryptographyClient::Encrypt(
+    EncryptParameters const& parameters,
+    Azure::Core::Context const& context)
+{
+  if (m_provider == nullptr)
+  {
+    // Try to init a local crypto provider after getting the key from the server.
+    // If the local provider can't be created, the remote client is used as provider.
+    Initialize(KeyOperation::Encrypt.ToString(), context);
+  }
+
+  // Default result has empty values.
+  EncryptResult result;
+
+  // m_provider can be local or remote, depending on how it was init.
+  if (m_provider->SupportsOperation(KeyOperation::Encrypt))
+  {
+    try
+    {
+      result = m_provider->Encrypt(parameters, context);
+    }
+    catch (std::exception const&)
+    {
+      // If provider supports remote, otherwise re-throw
+      if (!m_provider->CanRemote())
+      {
+        throw;
+      }
+    }
+
+    if (result.Ciphertext.empty())
+    {
+      ThrowIfLocalOnly(KeyOperation::Encrypt.ToString());
+
+      result = m_remoteProvider->Encrypt(parameters, context);
+    }
+  }
+
+  return result;
+}
+
+DecryptResult CryptographyClient::Decrypt(
+    DecryptParameters const& parameters,
+    Azure::Core::Context const& context)
+{
+  if (m_provider == nullptr)
+  {
```

Comment:
Because the CryptoClient can be constructed directly from a keyCient or from a JsonWebKey.

When constructed from a JsonWebKey, the crypto client can be initialized as it is constructed because the key material is already present.

But when the crypto client is constructed from a key client, they crypto client still needs to fetch the key material from the server to init a local provider.